### ISSUE TITLE
questions about how to use it

### ISSUE THREAD

**Issue by fangli333** [kuldeepbrd1/diff-3dgs-cam-grad#1]
Hi,

Thank you for your work. I think your work is really helpful to my project. I saw your work in the repo of 3dgs, and I want to ask did you proved that your code work can work? I noticed that you let others to check if your code is reasonable five days ago. So I am not sure whether I can directly use your code in my project. Btw, I also want to ask if use your code, what should i change in 3d-gs code. Thank you.

Best,
Kevin

**Comment by fangli333** [comment]
one more question is that did you take all the fov things and viewmatrix and projmatrix as gradient true?

**Comment by kuldeepbrd1** [comment]
@fangli333 Sorry, I have not been able to verify that this works as expected. I'm not sure yet if this is a problem of derivation or implementation mistake. 

I have been trying to verify with SLAM use case (like https://github.com/spla-tam/SplaTAM), but there seems to be an issue with quaternion gradients on this [quaternion gradient branch](https://github.com/kuldeepbrd1/diff-3dgs-cam-grad/tree/quaternion_grads).

If you want to try it out, you just need to supply the pose in this call
```shell
rasterize_gaussians(
    means3D,
    means2D,
    sh,
    colors_precomp,
    opacities,
    scales,
    rotations,
    cov3Ds_precomp,
    raster_settings,
    pose_1x7,
)
```

`pose_1x7` is `[ x, y, z, qw, qx, qy, qz]` which is `Tensor` with grads on. The backward method [here](https://github.com/kuldeepbrd1/diff-3dgs-cam-grad/blob/4311fa3ae04e5f89acd6bfbaedd319e7d9f16195/diff_gaussian_rasterization/__init__.py#L187) will provide the gradients back to this parameter.


You need to add this to the appropriate optimizer `param_groups` and assign a learning rate. Other than that, I think you can keep grads off for everything else (not in `param_groups` of the optimizer).

**Comment by fangli333** [comment]
what do you mean by issue with quaternion? There are some code bugs? Which branch should I try?

**Comment by kuldeepbrd1** [comment]
The latest work is on `quaternion_grads` [branch](https://github.com/kuldeepbrd1/diff-3dgs-cam-grad/tree/quaternion_grads). You can pull it if you'd like to try and investigate.

Yes, there seem to be bugs. I cannot definitively say where the error is in implementation atm.  I would not use the code as it is for any purpose, unless for debugging and investigating.

**Comment by fangli333** [comment]
why do you introduce a new variable? The original code has fov and transform matrix. I guess maybe changing the unlearnable variable to learnable will be easier. but the backward part might be hard due to so many learnable variable of camera parameters. How do you think of it?

**Comment by fangli333** [comment]
I only have some basic knowledge of CUDA language. So I am not sure whether my thoughts are right.

**Comment by kuldeepbrd1** [comment]
Hi @fangli333 

I believe you are asking about introducing `pose_1x7` variable.

There are a few things here:
1. What we wish to do is to provide an explicit gradient for a tensor of our choice, in our case the view matrix or camera pose matrix. Because autograd in pytorch (for instance as is used in SplatAM) scales poorly with the number of gaussians. To do this we need to assign a gradient value to the pose/viewmatrix tensor. 

2. For such purposes, we need to implement a custom backward method that tells pytorch, what gradients to use in backpropagation for the tensors used in the forward propagation.  

3. In our case we want a custom `backward` implementation, where we return explicit gradients *with respect to the Loss*. The gradients w.r.t pose are computed efficiently inside the cuda code. See how this comes together in [\_\_init\_\_.py](https://github.com/kuldeepbrd1/diff-3dgs-cam-grad/blob/quaternion_grads/diff_gaussian_rasterization/__init__.py)

4. In the forward method, we save all tensors that require gradients inside `ctx`. During backward, we need to supply gradient for each of these. Pytorch will then assign it to the `grad` property of those tensors and take them into the computational graph. 



Finally, to your question - the reason why I use `pose_1x7` is purely because, I wish to calculate gradients for each of the 7 elements of the transformatio (3- translation and 4 rotation). One could also do it with respect all 16 elements of the full view matrix. But, that has a lot more redundant parameters that are optimized independently. Since you cannot enforce orthogonal constraint to the rotation matrix inside it during optimization, the optimization tends to be not so well-behaved. So, I prefer to use quaternion representation for the sake of optimization.

In the 3dgs implementation, they do not compute gradients for camera parameters and view matrix. It is only used for forward rendering (no gradients and no optimization). 

In my implementation, since I use 1x7 pose vector for optimization, what I'd do is that I'll always derive the view matrix from this pose vector by converting the quaternion to a rotation matrix and then supplying it inside `rasterization_settings`. In this case, what we can do is that the view matrix we supply to the renderer can be derived from this pose vector. 

See more info on extending Pytorch for custom modules [here](https://pytorch.org/docs/stable/notes/extending.html).

I hope this helps

**Comment by fangli333** [comment]
Thank you for your explanantion. I think your operation is reasonable, but unfortunately it has some bugs.....
